jxl/src/features/spline.rs

```diff
@@ -270,11 +347,213 @@ pub struct Splines {
     segment_y_start: Vec<usize>,
 }
 
+struct ExtendedPoints<'a> {
+    first_element: Point,
+    last_element: Point,
+    backend: &'a Vec<Point>,
+}
+
+impl ExtendedPoints<'_> {
+    fn len(&self) -> usize {
+        self.backend.len() + 2
+    }
+    fn iter(&self) -> ExtendedPointsIterator {
+        ExtendedPointsIterator {
+            index: 0,
+            backend: self,
+        }
+    }
+}
+
+impl Index<usize> for ExtendedPoints<'_> {
+    type Output = Point;
+    fn index(&self, index: usize) -> &Self::Output {
+        if index == 0 {
+            &self.first_element
+        } else if index == self.len() - 1 {
+            &self.last_element
+        } else {
+            &self.backend[index - 1]
+        }
+    }
+}
+
+struct ExtendedPointsIterator<'a> {
+    index: usize,
+    backend: &'a ExtendedPoints<'a>,
+}
+
+impl Iterator for ExtendedPointsIterator<'_> {
+    type Item = Point;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.index < self.backend.len() {
+            let result = Some(self.backend[self.index]);
+            self.index += 1;
+            result
+        } else {
+            None
+        }
+    }
+}
+
+fn draw_centripetal_catmull_rom_spline(points: &Vec<Point>) -> Vec<Point> {
+    if points.is_empty() {
+        return [].to_vec();
+    }
+    if points.len() == 1 {
+        return vec![points[0]];
+    }
+    const NUM_POINTS: usize = 16;
+    // Create a view of points with one prepended and one appended point.
+    let extended_points = ExtendedPoints {
```

Comment:
While this works, it doesn't seem to me like the most idiomatic way to do this -- what about `iter::once(first).chain(points.iter()).chain(iter::once(last)).scan()`, for example?
I have not thought too much about the other uses of extended_points below, but ISTM that they'd be reasonable to express as operations on iterators too, avoiding the need for ExtendedPoints.

Reply:
The extended points list is indexed a few steps in front of the iterator all over the place, so a regular iterator wouldn't be enough.

What do you think is a reasonable way to avoid copying everything but still mapping index 0 to first_element and index len-1 to last_element?

Reply:
I would not use indexing in the first place. All indexing accesses contiguous positions, so it should be doable with iterators without major issues, which likely results in more efficient and more idiomatic code.

Reply:
Done, is this better?